### PR TITLE
fileservice: replace CacheStats with perfcounter 

### PR DESCRIPTION
**Pull request by reusee**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:

fileservice: optimize counter updating

fileservice: remove CacheStats

fileservice: make counters in context maps

fileservice: add CachingFileService.CacheCounter

### CODE REVIEW

**Review by fengttt — CHANGES_REQUESTED** [review]
please ask @arjunsk to review this code.

**Review by arjunsk — APPROVED** [review]

File: cmd/mo-service/config.go [hunk 1]
```diff
@@ -165,6 +166,35 @@ func (c *Config) createFileService(defaultName string) (*fileservice.FileService
 			return nil, err
 		}
 		services = append(services, service)
+
+		// cache stats
+		if fs, ok := service.(fileservice.CachingFileService); ok {
+			go func() {
+				for range time.NewTicker(time.Second * 10).C {
+					counter := fs.CacheCounter()
+
```
Comment (arjunsk) [inline cmd/mo-service/config.go:175]: I think @fengttt  doesn't want a goroutine per CounterCollection.

According to him, there will be several CounterCollections defined across CN, DN, FileService, LogTail, etc.

He wants one background goroutine to periodically collect the Values from all these CouterCollections and log them one by one.

There will be approx 1000 counters in total.

Definitions:
- counter is "S3ListObjects atomic.Int64"
- counterCollection is "type Counter struct"

**Review by arjunsk — COMMENTED** [review]
I think I will move this Logging logic to metrics pkg.

File: cmd/mo-service/config.go (hunk 1 above)
Comment (reusee) [reply cmd/mo-service/config.go]: Yes, I've changed to iterate all fs instances in one goroutine.
Just replace the fmt.Printf with metrics exporter in your following PR.

**Review by nnsgmsone — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

File: pkg/fileservice/mem_cache.go [hunk 2]
```diff
@@ -82,12 +91,14 @@ func (m *MemCache) Read(
 	_, span := trace.Start(ctx, "MemCache.Read")
```
Comment (xzxiong) [inline pkg/fileservice/mem_cache.go:91]: you can remove the trace on “S3FS.Read” and “MemCache.Read”

**Review by xzxiong — APPROVED** [review]

File: pkg/fileservice/mem_cache.go (hunk 2 above)
Comment (reusee) [reply pkg/fileservice/mem_cache.go]: Done

**Review by cnutshell — APPROVED** [review]

**Review by LeftHandCold — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by arjunsk — APPROVED** [review]

File: cmd/mo-service/config.go [hunk 3]
```diff
@@ -153,20 +156,61 @@ func (c *Config) validate() error {
 func (c *Config) createFileService(defaultName string) (*fileservice.FileServices, error) {
 	// create all services
 	services := make([]fileservice.FileService, 0, len(c.FileServices))
+
+	type perfCounterInfo struct {
+		fsName      string
+		perfCounter *perfcounter.Counter
```
Comment (fengttt) [inline cmd/mo-service/config.go:162]: does this really need to be a pointer?   can it be a struct member?

File: cmd/mo-service/config.go [hunk 4]
```diff
@@ -153,20 +156,61 @@ func (c *Config) validate() error {
...
+
 	}
 
+	// cache stats
+	//TODO add metrics exporter
+	go func() {
+		for range time.NewTicker(time.Second * 10).C {
+			for _, info := range perfCounterInfos {
+
+				reads := atomic.LoadInt64(&info.perfCounter.CacheRead)
```
Comment (fengttt) [inline cmd/mo-service/config.go:192]: dont you need to reset counter for next ticker period?

File: pkg/fileservice/mem_cache.go [hunk 5]
```diff
@@ -79,15 +87,15 @@ func (m *MemCache) Read(
...
-	defer span.End()
 
-	numHit := 0
+	var numHit, numRead int64
 	defer func() {
-		if m.stats != nil {
-			atomic.AddInt64(&m.stats.NumRead, int64(len(vector.Entries)))
-			atomic.AddInt64(&m.stats.NumHit, int64(numHit))
-		}
+		perfcounter.Update(ctx, func(c *perfcounter.Counter) {
```
Comment (fengttt) [inline pkg/fileservice/mem_cache.go:93]: hmm, this call is still kind of clumsy.    anyway to simplify this call?

File: cmd/mo-service/config.go (hunk 4 above)
Comment (reusee) [reply cmd/mo-service/config.go:192]: No, it's a global counter.
This part will be replaced with Arjun's collector.

File: pkg/fileservice/mem_cache.go (hunk 5 above)
Comment (reusee) [reply pkg/fileservice/mem_cache.go:93]: We have to update counters in the context and counters passed in NewLocalFS and NewS3FS, and we have to specify which fields to increase.
Maybe we can use atomic.Int64 and the code will be c.CacheRead.Add(1).

**Review by sukki37 — APPROVED** [review]

File: cmd/mo-service/config.go (hunk 3 above)
Comment (reusee) [reply cmd/mo-service/config.go]: Done

File: pkg/fileservice/mem_cache.go (hunk 5 above)
Comment (reusee) [reply pkg/fileservice/mem_cache.go:92]: Changed to use atomic.Int64

**Review by fengttt — APPROVED** [review]
I am fine with the PR.    Some minors.

Lets give the general solution work to Arjun.

**Pull request merged**